packages/dds/tree/src/feature-libraries/modular-schema/modularChangeFamily.ts

```diff
@@ -218,106 +224,270 @@ export class ModularChangeFamily
+		const composedNodeAliases: ChangeAtomIdMap<NodeId> = mergeNestedMaps(
+			change1.nodeAliases,
+			change2.nodeAliases,
+		);
 
-		crossFieldTable.nodeIdPairs.length = 0;
+		this.composeDiscoveredFields(
+			change1,
+			change2,
+			crossFieldTable,
+			composedFields,
+			composedNodeChanges,
+			composedNodeToParent,
+			composedNodeAliases,
+			genId,
+			revisionMetadata,
+		);
 
 		while (crossFieldTable.invalidatedFields.size > 0) {
 			const fieldsToUpdate = crossFieldTable.invalidatedFields;
 			crossFieldTable.invalidatedFields = new Set();
 			for (const fieldChange of fieldsToUpdate) {
-				const context = crossFieldTable.fieldToContext.get(fieldChange);
-				assert(
-					context !== undefined,
-					0x8cc /* Should have context for every invalidated field */,
+				this.composeInvalidatedField(
+					change1,
+					change2,
+					composedNodeChanges,
+					fieldChange,
+					crossFieldTable,
+					genId,
+					revisionMetadata,
 				);
-				const { change1: fieldChange1, change2: fieldChange2, composedChange } = context;
-
-				const rebaser = getChangeHandler(this.fieldKinds, composedChange.fieldKind).rebaser;
-				const composeNodes = (
-					child1: NodeId | undefined,
-					child2: NodeId | undefined,
-				): NodeId => {
-					if (
-						child2 !== undefined &&
-						!nestedSetContains(crossFieldTable.nodeIds, child2.revision, child2.localId)
-					) {
-						crossFieldTable.nodeIdPairs.push([child1, child2]);
-						if (child1 !== undefined && child2 !== undefined) {
-							addToNestedSet(crossFieldTable.nodeIds, child2.revision, child2.localId);
-						}
-					}
-					return child1 ?? child2 ?? fail("Should not compose two undefined nodes");
-				};
 
-				const amendedChange = rebaser.compose(
-					fieldChange1,
-					fieldChange2,
-					composeNodes,
+				// Process any newly discovered fields.
+				this.composeDiscoveredFields(
+					change1,
+					change2,
+					crossFieldTable,
+					composedFields,
+					composedNodeChanges,
+					composedNodeToParent,
+					composedNodeAliases,
 					genId,
-					newCrossFieldManager(crossFieldTable, fieldChange, false),
 					revisionMetadata,
 				);
-				composedChange.change = brand(amendedChange);
-
-				// Process any newly discovered nodes.
-				for (const [taggedId1, taggedId2] of crossFieldTable.nodeIdPairs) {
-					this.composeNodesById(
-						change1.change.nodeChanges,
-						change2.change.nodeChanges,
-						composedNodeChanges,
-						taggedId1,
-						taggedId2,
-						genId,
-						crossFieldTable,
-						revisionMetadata,
-					);
-				}
-
-				crossFieldTable.nodeIdPairs.length = 0;
 			}
 		}
 
-		const { allBuilds, allDestroys, allRefreshers } = composeBuildsDestroysAndRefreshers([
-			change1,
-			change2,
-		]);
+		return {
+			fieldChanges: composedFields,
+			nodeChanges: composedNodeChanges,
+			nodeToParent: composedNodeToParent,
+			nodeAliases: composedNodeAliases,
+			crossFieldKeys: brand(mergeBTrees(change1.crossFieldKeys, change2.crossFieldKeys)),
```

Comment:
Shouldn't some cross field keys be dropped (e.g., when composing moves from A to B and B to C)?

Reply:
It's true that in some cases we could remove (and potentially have to later re-add) keys during compose or rebase, but I don't think there's any harm in leaving unnecessary keys.

It's also worth noting that a field handler might want to move keys between fields during compose, but this is currently not supported since at present none of our fields need that feature.

Reply:
Please add comments